### DISCUSSION TITLE
LightningCLI: callbacks based on model

### DISCUSSION THREAD

**Discussion by adamjstewart** [Lightning-AI/pytorch-lightning#18480] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule] [answered]
I'm trying to use `LightningCLI` to replace TorchGeo's [train.py](https://github.com/microsoft/torchgeo/blob/b30a8b939325be2a73ffbd61332e506966700393/train.py) script. I'm absolutely loving it so far, but have encountered one thing I'm not sure how to translate. The performance metric we monitor in our ModelCheckpoint/EarlyStopping callbacks depends on which model we select (e.g., classification: 'val_loss', object detection: 'val_map', SSL: 'train_loss'). What would be the most idiomatic way to handle this?

My first thought was to override `LightningCLI.instantiate_trainer` so that we can add callbacks based on parsed model config. Is there a cleaner way to do this? Ideally it would be possible for the user to override this but it's not a requirement.

**Comment by tshu-w** [comment, marked as answer]
Maybe you can config callbacks for model directly through [`LightningModule.configure_callbacks`](https://lightning.ai/docs/pytorch/stable/api/lightning.pytorch.core.LightningModule.html#lightning.pytorch.core.LightningModule.configure_callbacks)

**Reply by adamjstewart** [reply]
I don't know how I've never discovered this before but this makes so much more sense, thanks!
